account/recovery.go

```diff
@@ -220,16 +379,121 @@ func recoverInitialState(ctx context.Context, cfg RecoveryConfig) ([]*Account,
 func updateAccountStates(cfg RecoveryConfig, client *rpcclient.Client,
 	accounts []*Account) ([]*Account, error) {
 
-	// TODO (positiveblue): update account states
+	for i := cfg.FirstBlock; i < uint32(cfg.CurrentBlockHeight); i++ {
+		blockHeight := i
+
+		blockHash, err := client.GetBlockHash(int64(i))
+		if err != nil {
+			return nil, err
+		}
+
+		block, err := client.GetBlock(blockHash)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, tx := range block.Transactions {
+			for idx, acc := range accounts {
+				// It's safe to ignore accounts that are in a
+				// terminal state.
+				if acc.State == StateClosed {
+					continue
+				}
+				found := poolscript.IncludesPreviousOutPoint(
+					tx, acc.OutPoint,
+				)
+
+				if !found {
+					continue
+				}
+
+				newAcc, err := findAccountUpdate(
+					cfg, acc, blockHeight, tx,
+				)
+				if err != nil {
+					log.Debugf("unable to find account "+
+						"update: %v", err)
+					continue
+				}
+
+				accounts[idx] = newAcc
+			}
+		}
+	}
+	return accounts, nil
+}
+
+// findAccountUpdate tries to find the new account values after an update.
+func findAccountUpdate(cfg RecoveryConfig, acc *Account, blockHeight uint32,
+	tx *wire.MsgTx) (*Account, error) {
+
+	newAcc := acc.Copy()
+	newAcc.BatchKey = poolscript.IncrementKey(newAcc.BatchKey)
+
+	idx, ok := matchScript(newAcc, newAcc.Expiry, tx)
+	if ok {
+		newAcc.Value = btcutil.Amount(tx.TxOut[idx].Value)
+		// TODO (positiveblue): change 0 for dust limit.
+		if newAcc.Value == 0 {
+			newAcc.State = StateClosed
+		}
+
+		newAcc.OutPoint = wire.OutPoint{
+			Hash:  tx.TxHash(),
+			Index: idx,
+		}
+		newAcc.LatestTx = tx
+		newAcc.HeightHint = blockHeight
+
+		return newAcc, nil
+	}
+
+	// If the update included a new expiration date we need to brute force
+	// our new expiration date again.
+	for height := cfg.FirstBlock; height <= cfg.LastBlock; height++ {
+		idx, ok := matchScript(newAcc, height, tx)
+		if !ok {
+			continue
+		}
+
+		newAcc.Expiry = height
+		newAcc.Value = btcutil.Amount(tx.TxOut[idx].Value)
+		// TODO (positiveblue): change 0 for dust limit.
+		if newAcc.Value == 0 {
+			newAcc.State = StateClosed
+		}
+		newAcc.OutPoint = wire.OutPoint{
+			Hash:  tx.TxHash(),
+			Index: idx,
+		}
+		newAcc.HeightHint = blockHeight
+		newAcc.LatestTx = tx
+
+		return newAcc, nil
+	}
+
+	return nil, fmt.Errorf("account update not found")
+}
+
+// matchScript creates a new Account script and tries to locate the output in a
+// transaction.
+func matchScript(acc *Account, expiry uint32, tx *wire.MsgTx) (uint32, bool) {
```

Comment:
nit: since this is now only used by `findAccountUpdate`, we could use the helper here too to speed things up a bit.